src/System.IO.Pipelines/PipelineReaderWriter.cs

```diff
@@ -366,19 +367,21 @@ public void AdvanceReader(ReadCursor consumed, ReadCursor examined)
             // Reading commit head shared with writer
             lock (_sync)
             {
-                if (!examined.IsDefault &&
+                if ((!examined.IsDefault &&
                     examined.Segment == _commitHead &&
                     examined.Index == _commitHeadIndex &&
-                    Reading.Status == TaskStatus.WaitingForActivation)
+                    Reading.Status == TaskStatus.WaitingForActivation) || _readingIsCancelled)
```

Comment:
I don't think you should have merged this. I'm pretty sure this is the opposite of what you want. Here's what could happen:

1. Data arrives, ReadAsync() completes because of the new data. (i.e. _awaitableState is called invoking consumer code).
2. Before Advance() is called, CancelPendingRead() is called (probably on another thread, but could be the consuming thread).
3. Advance() is called, but it didn't examine up to the commit head.
4. The next call to await ReadAsync(), instead of completing immediately like one would expect, hangs until more data arrives (which may never happen). This is because `_readingIsCancelled` was true during the last call to Advance() causing the `_awaitableState` to be transitioned to `_awaitableIsNotCompleted`. Now IReadableBufferAwaiter.IsCompleted is false.

You should add a test for calling CancelPendingRead() between ReadAsync() and Advance().